File: python/mirascope/llm/content/audio.py
"""The `Audio` content class."""

import base64
from dataclasses import dataclass
from pathlib import Path
from typing import Literal, get_args

import httpx

AudioMimeType = Literal[
    "audio/wav",
    "audio/mp3",
    "audio/aiff",
    "audio/aac",
    "audio/ogg",
    "audio/flac",
]  # TODO: add e2e tests for every supported type

MIME_TYPES = get_args(AudioMimeType)

# Maximum audio size in bytes (25MB)
MAX_AUDIO_SIZE = 25 * 1024 * 1024


def infer_audio_type(audio_data: bytes) -> AudioMimeType:
    """Get the MIME type of an audio file from its raw bytes.

    Raises:
        ValueError: If the audio type cannot be determined or data is too small
    """
    if len(audio_data) < 12:
        raise ValueError("Audio data too small to determine type (minimum 12 bytes)")

    if audio_data.startswith(b"RIFF") and audio_data[8:12] == b"WAVE":
        return "audio/wav"
    elif audio_data.startswith(b"ID3") or audio_data.startswith(b"\xff\xfb"):
        return "audio/mp3"
    elif audio_data.startswith(b"FORM") and audio_data[8:12] == b"AIFF":
        return "audio/aiff"
    elif audio_data.startswith(b"\xff\xf1") or audio_data.startswith(b"\xff\xf9"):
        return "audio/aac"
    elif audio_data.startswith(b"OggS"):
        return "audio/ogg"
    elif audio_data.startswith(b"fLaC"):
        return "audio/flac"
    raise ValueError("Unsupported audio type")


@dataclass(kw_only=True)
class Base64AudioSource:
    """Audio data represented as a base64 encoded string."""

    type: Literal["base64_audio_source"]

    data: str
    """The audio data, as a base64 encoded string."""

    mime_type: AudioMimeType
    """The mime type of the audio (e.g. audio/mp3)."""


def _process_audio_bytes(data: bytes, max_size: int) -> Base64AudioSource:
    """Validate and process audio bytes into a Base64AudioSource.

    Args:
        data: Raw audio bytes
        max_size: Maximum allowed size in bytes

    Returns:
        A Base64AudioSource with validated and encoded data

    Raises:
        ValueError: If data size exceeds max_size
    """
    size = len(data)
    if size > max_size:
        raise ValueError(
            f"Audio size ({size} bytes) exceeds maximum allowed size ({max_size} bytes)"
        )

    mime_type = infer_audio_type(data)
    encoded_data = base64.b64encode(data).decode("utf-8")
    return Base64AudioSource(
        type="base64_audio_source",
        data=encoded_data,
        mime_type=mime_type,
    )


@dataclass(kw_only=True)
class Audio:
    """Audio content for a message.

    Audio can be included in messages for voice or sound-based interactions.
    """

    type: Literal["audio"] = "audio"

    source: Base64AudioSource

    @classmethod
    def download(cls, url: str, *, max_size: int = MAX_AUDIO_SIZE) -> "Audio":
        """Download and encode an audio file from a URL.

        Args:
            url: The URL of the audio file to download
            max_size: Maximum allowed audio size in bytes (default: 25MB)

        Returns:
            An `Audio` with a `Base64AudioSource`

        Raises:
            ValueError: If the downloaded audio exceeds max_size
        """
        response = httpx.get(url, follow_redirects=True)
        response.raise_for_status()
        return cls(source=_process_audio_bytes(response.content, max_size))

    @classmethod
    async def download_async(
        cls, url: str, *, max_size: int = MAX_AUDIO_SIZE
    ) -> "Audio":
        """Asynchronously download and encode an audio file from a URL.

        Args:
            url: The URL of the audio file to download
            max_size: Maximum allowed audio size in bytes (default: 25MB)

        Returns:
            An `Audio` with a `Base64AudioSource`

        Raises:
            ValueError: If the downloaded audio exceeds max_size
        """
        async with httpx.AsyncClient() as client:
            response = await client.get(url, follow_redirects=True)
            response.raise_for_status()
            return cls(source=_process_audio_bytes(response.content, max_size))

    @classmethod
    def from_file(cls, file_path: str, *, max_size: int = MAX_AUDIO_SIZE) -> "Audio":
        """Create an `Audio` from a file path.

        Args:
            file_path: Path to the audio file
            max_size: Maximum allowed audio size in bytes (default: 25MB)

        Raises:
            FileNotFoundError: If the file does not exist
            ValueError: If the file size exceeds max_size
        """
        path = Path(file_path)
        file_size = path.stat().st_size
        if file_size > max_size:
            raise ValueError(
                f"Audio file size ({file_size} bytes) exceeds maximum allowed size ({max_size} bytes)"
            )
        with open(path, "rb") as f:
            audio_bytes = f.read()
        return cls(source=_process_audio_bytes(audio_bytes, max_size))

    @classmethod
    def from_bytes(cls, data: bytes, *, max_size: int = MAX_AUDIO_SIZE) -> "Audio":
        """Create an `Audio` from raw bytes.

        Args:
            data: Raw audio bytes
            max_size: Maximum allowed audio size in bytes (default: 25MB)

        Raises:
            ValueError: If the data size exceeds max_size
        """
        return cls(source=_process_audio_bytes(data, max_size))

File: typescript/src/llm/content/audio.ts
/**
 * Supported audio MIME types.
 */
export type AudioMimeType =
  | 'audio/wav'
  | 'audio/mp3'
  | 'audio/aiff'
  | 'audio/aac'
  | 'audio/ogg'
  | 'audio/flac';

/**
 * Audio data encoded as base64.
 */
export type Base64AudioSource = {
  readonly type: 'base64_audio_source';

  /** The audio data, as a base64 encoded string. */
  readonly data: string;

  /** The mime type of the audio (e.g. audio/mp3). */
  readonly mimeType: AudioMimeType;
};

/**
 * Audio content for a message.
 *
 * Audio can be included in user messages for multimodal models that support it.
 */
export type Audio = {
  readonly type: 'audio';

  readonly source: Base64AudioSource;
};

/** Maximum audio size in bytes (25MB) */
export const MAX_AUDIO_SIZE = 25 * 1024 * 1024;

/**
 * Infer the MIME type of audio from its magic bytes.
 *
 * @throws Error if the audio type cannot be determined
 */
export function inferAudioType(data: Uint8Array): AudioMimeType {
  if (data.length < 12) {
    throw new Error(
      'Audio data too small to determine type (minimum 12 bytes)'
    );
  }

  // WAV: starts with RIFF....WAVE
  if (
    data[0] === 0x52 &&
    data[1] === 0x49 &&
    data[2] === 0x46 &&
    data[3] === 0x46 &&
    data[8] === 0x57 &&
    data[9] === 0x41 &&
    data[10] === 0x56 &&
    data[11] === 0x45
  ) {
    return 'audio/wav';
  }

  // MP3: starts with ID3 or 0xFF 0xFB (frame sync)
  if (
    (data[0] === 0x49 && data[1] === 0x44 && data[2] === 0x33) ||
    (data[0] === 0xff &&
      (data[1] === 0xfb ||
        data[1] === 0xfa ||
        data[1] === 0xf3 ||
        data[1] === 0xf2))
  ) {
    return 'audio/mp3';
  }

  // AIFF: starts with FORM....AIFF
  if (
    data[0] === 0x46 &&
    data[1] === 0x4f &&
    data[2] === 0x52 &&
    data[3] === 0x4d &&
    data[8] === 0x41 &&
    data[9] === 0x49 &&
    data[10] === 0x46 &&
    data[11] === 0x46
  ) {
    return 'audio/aiff';
  }

  // OGG: starts with OggS
  if (
    data[0] === 0x4f &&
    data[1] === 0x67 &&
    data[2] === 0x67 &&
    data[3] === 0x53
  ) {
    return 'audio/ogg';
  }

  // FLAC: starts with fLaC
  if (
    data[0] === 0x66 &&
    data[1] === 0x4c &&
    data[2] === 0x61 &&
    data[3] === 0x43
  ) {
    return 'audio/flac';
  }

  // AAC: starts with ADTS sync word 0xFFF (first 12 bits)
  // Safe to use ! - we validated length >= 12 above
  if (data[0] === 0xff && (data[1]! & 0xf0) === 0xf0) {
    return 'audio/aac';
  }

  throw new Error('Unsupported audio type');
}

/**
 * Convert a Uint8Array to a base64 string.
 */
function uint8ArrayToBase64(data: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < data.length; i++) {
    binary += String.fromCharCode(data[i]!);
  }
  return btoa(binary);
}

/**
 * Factory methods for creating Audio instances.
 */
export const Audio = {
  /**
   * Download audio from a URL and encode as base64.
   *
   * @throws Error if download fails or audio exceeds maxSize
   */
  download: async (url: string, maxSize = MAX_AUDIO_SIZE): Promise<Audio> => {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(
        `Failed to download audio: ${response.status} ${response.statusText}`
      );
    }
    const buffer = await response.arrayBuffer();
    return Audio.fromBytes(new Uint8Array(buffer), maxSize);
  },

  /**
   * Create Audio from raw bytes.
   *
   * @throws Error if data exceeds maxSize or type cannot be inferred
   */
  fromBytes: (data: Uint8Array, maxSize = MAX_AUDIO_SIZE): Audio => {
    if (data.length > maxSize) {
      throw new Error(
        `Audio size (${data.length} bytes) exceeds maximum (${maxSize} bytes)`
      );
    }
    const mimeType = inferAudioType(data);
    const base64 = uint8ArrayToBase64(data);
    return {
      type: 'audio',
      source: { type: 'base64_audio_source', data: base64, mimeType },
    };
  },
};
